client/me/account/main.jsx

```diff
@@ -84,25 +82,58 @@ const Account = React.createClass( {
 		debug( this.constructor.displayName + ' component is unmounting.' );
 	},
 
-	updateLanguage() {
-		let valueLink = this.valueLink( 'language' );
+	getUserSetting( settingName ) {
+		return this.props.userSettings.getSetting( settingName );
+	},
 
-		valueLink.requestChange = ( value ) => {
-			const originalLanguage = this.props.userSettings.getOriginalSetting( 'language' );
+	updateUserSetting( settingName, value ) {
+		this.props.userSettings.updateSetting( settingName, value );
+	},
 
-			this.props.userSettings.updateSetting( 'language', value );
-			if ( value !== originalLanguage ) {
-				this.setState( { redirect: '/me/account' } );
-			} else {
-				this.setState( { redirect: false } );
-			}
-		};
+	updateUserSettingInput( settingName ) {
+		return event => this.updateUserSetting( settingName, event.target.value );
+	},
 
-		return valueLink;
+	updateUserSettingCheckbox( settingName ) {
+		return event => this.updateUserSetting( settingName, event.target.checked );
+	},
+
+	updateLanguage( event ) {
+		const { value } = event.target;
+		const originalLanguage = this.props.userSettings.getOriginalSetting( 'language' );
+
+		this.updateUserSetting( 'language', value );
+		if ( value !== originalLanguage ) {
+			this.setState( { redirect: '/me/account' } );
+		} else {
+			this.setState( { redirect: false } );
+		}
+	},
+
+	getEmailAddress() {
+		return this.hasPendingEmailChange()
+			? this.getUserSetting( 'new_user_email' )
+			: this.getUserSetting( 'user_email' );
+	},
+
+	updateEmailAddress( event ) {
+		const { value } = event.target;
+		if ( '' === value ) {
+			this.setState( { emailValidationError: 'empty' } );
+		} else if ( ! emailValidator.validate( value ) ) {
+			this.setState( { emailValidationError: 'invalid' } );
+		} else {
+			this.setState( { emailValidationError: false } );
+		}
```

Comment:
the `false` in my original concoction wasn't even necessary because if the first two conditions fail the result will be `false` but I left it in there to try and indicate more explicitly what would happen. that's really quite a difficult thing the way it's checking

```haskell
-- if only…
validationError = isEmpty `mplus` isInvalid
	where
		isEmpty = if "" == value then Just "empty" else Nothing
		isInvalid = if (not . validate) value then Just "invalid" else Nothing
```